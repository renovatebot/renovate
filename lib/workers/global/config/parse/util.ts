import { dequal } from 'dequal';
import { massageConfig } from '../../../../config/massage';
import { migrateConfig } from '../../../../config/migration';
import type { RenovateConfig } from '../../../../config/types';
import { validateConfig } from '../../../../config/validation';
import { logger } from '../../../../logger';

export async function migrateAndValidateConfig(
  config: RenovateConfig,
  configType: string,
): Promise<RenovateConfig> {
  const { isMigrated, migratedConfig } = migrateConfig(config);
  if (isMigrated) {
    logger.warn(
      { originalConfig: config, migratedConfig },
      `${configType} needs migrating`,
    );
  }
  const massagedConfig = massageConfig(migratedConfig);
  // log only if it's changed
  if (!dequal(migratedConfig, massagedConfig)) {
    logger.trace({ config: massagedConfig }, 'Post-massage config');
  }

  const { warnings, errors } = await validateConfig('global', massagedConfig);

  if (warnings.length) {
    logger.warn(
      { warnings },
      `Config validation warnings found in ${configType}`,
    );
  }
  if (errors.length) {
    logger.warn({ errors }, `Config validation errors found in ${configType}`);
  }

  return massagedConfig;
}
