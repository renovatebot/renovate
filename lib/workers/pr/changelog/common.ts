import { Release } from '../../../datasource';

export interface ChangeLogNotes {
  body?: string;
  id?: number;
  name?: string;
  tag?: string;
  url: string;
}

export interface ChangeLogChange {
  date: Date;
  message: string;
  sha: string;
}

export interface ChangeLogRelease {
  changes: ChangeLogChange[];
  compare: { url?: string };
  date: string | Date;
  releaseNotes?: ChangeLogNotes;
  version: string;
}

export interface ChangeLogProject {
  depName?: string;
  github: string;
  githubApiBaseURL?: string;
  githubBaseURL: string;
  repository: string;
}

export interface ChangeLogResult {
  hasReleaseNotes?: boolean;
  project: ChangeLogProject;
  versions: ChangeLogRelease[];
}

export interface ChangeLogConfig {
  depName: string;
  depType?: string;
  endpoint: string;
  fromVersion: string;
  manager?: string;
  releases: Release[];
  sourceUrl?: string;
  toVersion: string;
  versionScheme: string;
}
