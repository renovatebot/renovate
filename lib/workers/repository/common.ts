import { nameFromLevel } from 'bunyan';
import { getProblems } from '../../logger';

export function extractRepoProblems(
  repository: string | undefined,
): Set<string> {
  return new Set(
    getProblems()
      .filter(
        (problem) =>
          problem.repository === repository && !problem.artifactErrors,
      )
      .map(
        (problem) =>
          `${nameFromLevel[problem.level].toUpperCase()}: ${problem.msg}`,
      ),
  );
}
