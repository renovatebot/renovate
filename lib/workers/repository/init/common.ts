export type RepoConfigError = {
  validationError: string;
  validationMessage: string;
};

export interface RepoFileConfig {
  configFileName?: string;
  configFileParsed?: any;
  configFileParseError?: RepoConfigError;
}

export interface RepoInitConfig {
  defaultBranchSha?: string;
  repoConfig?: RepoFileConfig;
}
