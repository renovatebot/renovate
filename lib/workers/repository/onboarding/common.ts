import { configFileNames } from '../../../config/app-strings';
import type { RenovateConfig } from '../../../config/types';
import { logger } from '../../../logger';
import * as memCache from '../../../util/cache/memory';

export function defaultConfigFile(config: RenovateConfig): string {
  return configFileNames.includes(config.onboardingConfigFileName!)
    ? config.onboardingConfigFileName!
    : configFileNames[0];
}

export class OnboardingState {
  private static readonly cacheKey = 'OnboardingState';
  private static readonly skipKey = 'OnboardingStateValid';

  static get prUpdateRequested(): boolean {
    const updateRequested = !!memCache.get<boolean | undefined>(
      OnboardingState.cacheKey
    );
    logger.trace(
      { value: updateRequested },
      'Get OnboardingState.prUpdateRequested'
    );
    return updateRequested;
  }

  static set prUpdateRequested(value: boolean) {
    logger.trace({ value }, 'Set OnboardingState.prUpdateRequested');
    memCache.set(OnboardingState.cacheKey, value);
  }

  static get onboardingCacheValid(): boolean {
    const cacheValid = !!memCache.get<boolean | undefined>(
      OnboardingState.skipKey
    );
    logger.trace(
      { value: cacheValid },
      'Get OnboardingState.onboardingCacheValid'
    );
    return cacheValid;
  }

  static set onboardingCacheValid(value: boolean) {
    logger.trace({ value }, 'Set OnboardingState.onboardingCacheValid');
    memCache.set(OnboardingState.skipKey, value);
  }
}
