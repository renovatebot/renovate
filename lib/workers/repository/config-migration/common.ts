import type { RenovateConfig } from '../../../config/types';
import * as template from '../../../util/template';

const migrationBranchTemplate = '{{{branchPrefix}}}migrate-config';

export function getMigrationBranchName(config: RenovateConfig): string {
  return template.compile(migrationBranchTemplate, config);
}
