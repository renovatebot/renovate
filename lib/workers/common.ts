import { PrBodyUpgrade, PrBodyConfig } from './pr/body/common';
import { PackageDependency } from '../manager/common';
import { ChangeLogConfig } from './pr/changelog';
import { RenovateConfig } from '../config';
import { StabilityConfig, UnpublishableConfig } from './branch/status-checks';
import { CommitConfig } from './branch/commit';

export type PrUpgrade = PrBodyUpgrade &
  PackageDependency &
  ChangeLogConfig &
  PrBodyConfig & {
    manager?: string;
    hasUrls?: boolean;
    githubName?: string | undefined;
  };

export type BranchConfig = RenovateConfig &
  StabilityConfig &
  UnpublishableConfig &
  CommitConfig &
  PrBodyConfig<PrUpgrade>;
