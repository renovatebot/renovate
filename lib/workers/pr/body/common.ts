import { LookupUpdate } from '../../repository/process/lookup/common';
import { RenovateSharedConfig } from '../../../config';
import { Release, ReleaseResult } from '../../../datasource';

export type PrBodyUpgrade = LookupUpdate &
  Release &
  ReleaseResult & {
    prBodyNotes?: string[];
    depName?: string;
    depNameLinked?: string;
    references?: string;
    versionScheme: string;
    stabilityDays?: number;
  };

// TODO: Unify and optimize types and interfaces
export type PrBodyConfig<T extends PrBodyUpgrade = PrBodyUpgrade> = {
  upgrades: T[];
  prBanner?: string;
  prBodyTemplate: string;
  prBodyColumns: string[];
  prBodyDefinitions: Record<string, string>;
  updateType: string;
  isGroup?: boolean;
  isPin?: boolean;
  hasReleaseNotes?: boolean;
  global?: { prBanner?: string; prFooter?: string };
} & Partial<RenovateSharedConfig>;
