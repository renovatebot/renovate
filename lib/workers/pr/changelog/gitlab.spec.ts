import * as httpMock from '../../../../test/httpMock';
import { getName } from '../../../../test/util';
import { PLATFORM_TYPE_GITLAB } from '../../../constants/platforms';
import * as hostRules from '../../../util/host-rules';
import * as semverVersioning from '../../../versioning/semver';
import { BranchUpgradeConfig } from '../../common';
import { getChangeLogJSON } from '.';

jest.mock('../../../../lib/datasource/npm');

const upgrade: BranchUpgradeConfig = {
  branchName: undefined,
  endpoint: 'https://gitlab.com/api/v4/ ',
  depName: 'renovate',
  versioning: semverVersioning.id,
  fromVersion: '5.2.0',
  toVersion: '5.7.0',
  sourceUrl: 'https://gitlab.com/meno/dropzone/',
  releases: [
    // TODO: test gitRef
    { version: '5.2.0' },
    {
      version: '5.4.0',
      releaseTimestamp: '2018-08-24T14:23:00.000Z',
    },
    { version: '5.5.0', gitRef: 'eba303e91c930292198b2fc57040145682162a1b' },
    { version: '5.6.0', releaseTimestamp: '2020-02-13T15:37:00.000Z' },
    { version: '5.6.1' },
  ],
};

const baseUrl = 'https://gitlab.com/';

describe(getName(__filename), () => {
  describe('getChangeLogJSON', () => {
    beforeEach(() => {
      httpMock.setup();
      hostRules.clear();
      hostRules.add({
        hostType: PLATFORM_TYPE_GITLAB,
        baseUrl,
        token: 'abc',
      });
    });
    afterEach(() => {
      httpMock.reset();
    });
    it('returns null if @types', async () => {
      httpMock.scope(baseUrl);
      expect(
        await getChangeLogJSON({
          ...upgrade,
          fromVersion: null,
        })
      ).toBeNull();
      expect(httpMock.getTrace()).toBeEmpty();
    });
    it('returns null if fromVersion equals toVersion', async () => {
      httpMock.scope(baseUrl);
      expect(
        await getChangeLogJSON({
          ...upgrade,
          fromVersion: '1.0.0',
          toVersion: '1.0.0',
        })
      ).toBeNull();
      expect(httpMock.getTrace()).toBeEmpty();
    });
    it('skips invalid repos', async () => {
      expect(
        await getChangeLogJSON({
          ...upgrade,
          sourceUrl: 'https://gitlab.com/help',
        })
      ).toBeNull();
    });
    it('works without GitLab', async () => {
      expect(
        await getChangeLogJSON({
          ...upgrade,
        })
      ).toMatchSnapshot();
    });
    it('uses GitLab tags', async () => {
      httpMock
        .scope(baseUrl)
        .get('/api/v4/projects/meno%2Fdropzone/repository/tags')
        .reply(200, [
          { name: 'v5.2.0' },
          { name: 'v5.4.0' },
          { name: 'v5.5.0' },
          { name: 'v5.6.0' },
          { name: 'v5.6.1' },
          { name: 'v5.7.0' },
        ])
        .persist()
        .get('/api/v4/projects/meno/dropzone/repository/tree/')
        .reply(200, [])
        .persist()
        .get('/api/v4/projects/meno%2fdropzone/releases?per_page=100')
        .reply(200, []);
      expect(
        await getChangeLogJSON({
          ...upgrade,
        })
      ).toMatchSnapshot();
      expect(httpMock.getTrace()).toMatchSnapshot();
    });
    it('handles empty GitLab tags response', async () => {
      httpMock
        .scope(baseUrl)
        .get('/api/v4/projects/meno%2Fdropzone/repository/tags')
        .reply(200, [])
        .persist()
        .get('/api/v4/projects/meno/dropzone/repository/tree/')
        .reply(200, [])
        .persist()
        .get('/api/v4/projects/meno%2fdropzone/releases?per_page=100')
        .reply(200, []);
      expect(
        await getChangeLogJSON({
          ...upgrade,
        })
      ).toMatchSnapshot();
      expect(httpMock.getTrace()).toMatchSnapshot();
    });
    it('uses GitLab tags with error', async () => {
      httpMock
        .scope(baseUrl)
        .get('/api/v4/projects/meno%2Fdropzone/repository/tags')
        .replyWithError('Unknown GitLab Repo')
        .persist()
        .get('/api/v4/projects/meno/dropzone/repository/tree/')
        .reply(200, [])
        .persist()
        .get('/api/v4/projects/meno%2fdropzone/releases?per_page=100')
        .reply(200, []);
      expect(
        await getChangeLogJSON({
          ...upgrade,
        })
      ).toMatchSnapshot();
      expect(httpMock.getTrace()).toMatchSnapshot();
    });
    it('handles no sourceUrl', async () => {
      expect(
        await getChangeLogJSON({
          ...upgrade,
          sourceUrl: undefined,
        })
      ).toBeNull();
    });
    it('handles invalid sourceUrl', async () => {
      expect(
        await getChangeLogJSON({
          ...upgrade,
          sourceUrl: 'http://example.com',
        })
      ).toBeNull();
    });
    it('handles no releases', async () => {
      expect(
        await getChangeLogJSON({
          ...upgrade,
          releases: [],
        })
      ).toBeNull();
    });
    it('handles not enough releases', async () => {
      expect(
        await getChangeLogJSON({
          ...upgrade,
          releases: [{ version: '0.9.0' }],
        })
      ).toBeNull();
    });
    it('supports gitlab enterprise and gitlab enterprise changelog', async () => {
      hostRules.add({
        hostType: PLATFORM_TYPE_GITLAB,
        baseUrl: 'https://gitlab-enterprise.example.com/',
        token: 'abc',
      });
      process.env.GITHUB_ENDPOINT = '';
      expect(
        await getChangeLogJSON({
          ...upgrade,
          sourceUrl: 'https://gitlab-enterprise.example.com/meno/dropzone/',
          endpoint: 'https://gitlab-enterprise.example.com/',
        })
      ).toMatchSnapshot();
    });
  });
});
