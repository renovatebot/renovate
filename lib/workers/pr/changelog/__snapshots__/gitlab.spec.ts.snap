// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workers/pr/changelog/gitlab getChangeLogJSON handles empty GitLab tags response 1`] = `
Object {
  "hasReleaseNotes": false,
  "project": Object {
    "apiBaseUrl": "https://gitlab.com/api/v4/",
    "baseUrl": "https://gitlab.com/",
    "depName": "renovate",
    "gitlab": "meno/dropzone",
    "repository": "https://gitlab.com/meno/dropzone/",
  },
  "versions": Array [
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": undefined,
      "releaseNotes": null,
      "version": "5.6.1",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": "2020-02-13T15:37:00.000Z",
      "releaseNotes": null,
      "version": "5.6.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": undefined,
      "releaseNotes": null,
      "version": "5.5.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": "2018-08-24T14:23:00.000Z",
      "releaseNotes": null,
      "version": "5.4.0",
    },
  ],
}
`;

exports[`workers/pr/changelog/gitlab getChangeLogJSON handles empty GitLab tags response 2`] = `
Array [
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2Fdropzone/repository/tags",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
]
`;

exports[`workers/pr/changelog/gitlab getChangeLogJSON supports gitlab enterprise and gitlab enterprise changelog 1`] = `
Object {
  "hasReleaseNotes": false,
  "project": Object {
    "apiBaseUrl": "https://gitlab-enterprise.example.com/api/v4/",
    "baseUrl": "https://gitlab-enterprise.example.com/",
    "depName": "renovate",
    "gitlab": "meno/dropzone",
    "repository": "https://gitlab-enterprise.example.com/meno/dropzone/",
  },
  "versions": Array [
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": undefined,
      "releaseNotes": null,
      "version": "5.6.1",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": "2020-02-13T15:37:00.000Z",
      "releaseNotes": null,
      "version": "5.6.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": undefined,
      "releaseNotes": null,
      "version": "5.5.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": "2018-08-24T14:23:00.000Z",
      "releaseNotes": null,
      "version": "5.4.0",
    },
  ],
}
`;

exports[`workers/pr/changelog/gitlab getChangeLogJSON uses GitLab tags 1`] = `
Object {
  "hasReleaseNotes": true,
  "project": Object {
    "apiBaseUrl": "https://gitlab.com/api/v4/",
    "baseUrl": "https://gitlab.com/",
    "depName": "renovate",
    "gitlab": "meno/dropzone",
    "repository": "https://gitlab.com/meno/dropzone/",
  },
  "versions": Array [
    Object {
      "changes": Array [],
      "compare": Object {
        "url": "https://gitlab.com/meno/dropzone/compare/v5.6.0...v5.6.1",
      },
      "date": undefined,
      "releaseNotes": Object {
        "url": "https://gitlab.com/meno/dropzone/compare/v5.6.0...v5.6.1",
      },
      "version": "5.6.1",
    },
    Object {
      "changes": Array [],
      "compare": Object {
        "url": "https://gitlab.com/meno/dropzone/compare/v5.5.0...v5.6.0",
      },
      "date": "2020-02-13T15:37:00.000Z",
      "releaseNotes": Object {
        "url": "https://gitlab.com/meno/dropzone/compare/v5.5.0...v5.6.0",
      },
      "version": "5.6.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {
        "url": "https://gitlab.com/meno/dropzone/compare/v5.4.0...v5.5.0",
      },
      "date": undefined,
      "releaseNotes": Object {
        "url": "https://gitlab.com/meno/dropzone/compare/v5.4.0...v5.5.0",
      },
      "version": "5.5.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {
        "url": "https://gitlab.com/meno/dropzone/compare/v5.2.0...v5.4.0",
      },
      "date": "2018-08-24T14:23:00.000Z",
      "releaseNotes": Object {
        "url": "https://gitlab.com/meno/dropzone/compare/v5.2.0...v5.4.0",
      },
      "version": "5.4.0",
    },
  ],
}
`;

exports[`workers/pr/changelog/gitlab getChangeLogJSON uses GitLab tags 2`] = `
Array [
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2Fdropzone/repository/tags",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
]
`;

exports[`workers/pr/changelog/gitlab getChangeLogJSON uses GitLab tags with error 1`] = `
Object {
  "hasReleaseNotes": false,
  "project": Object {
    "apiBaseUrl": "https://gitlab.com/api/v4/",
    "baseUrl": "https://gitlab.com/",
    "depName": "renovate",
    "gitlab": "meno/dropzone",
    "repository": "https://gitlab.com/meno/dropzone/",
  },
  "versions": Array [
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": undefined,
      "releaseNotes": null,
      "version": "5.6.1",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": "2020-02-13T15:37:00.000Z",
      "releaseNotes": null,
      "version": "5.6.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": undefined,
      "releaseNotes": null,
      "version": "5.5.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": "2018-08-24T14:23:00.000Z",
      "releaseNotes": null,
      "version": "5.4.0",
    },
  ],
}
`;

exports[`workers/pr/changelog/gitlab getChangeLogJSON uses GitLab tags with error 2`] = `
Array [
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2Fdropzone/repository/tags",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno/dropzone/repository/tree/",
  },
  Object {
    "headers": Object {
      "accept": "application/json",
      "accept-encoding": "gzip, deflate",
      "host": "gitlab.com",
      "private-token": "abc",
      "user-agent": "https://github.com/renovatebot/renovate",
    },
    "method": "GET",
    "url": "https://gitlab.com/api/v4/projects/meno%2fdropzone/releases?per_page=100",
  },
]
`;

exports[`workers/pr/changelog/gitlab getChangeLogJSON works without GitLab 1`] = `
Object {
  "hasReleaseNotes": false,
  "project": Object {
    "apiBaseUrl": "https://gitlab.com/api/v4/",
    "baseUrl": "https://gitlab.com/",
    "depName": "renovate",
    "gitlab": "meno/dropzone",
    "repository": "https://gitlab.com/meno/dropzone/",
  },
  "versions": Array [
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": undefined,
      "releaseNotes": null,
      "version": "5.6.1",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": "2020-02-13T15:37:00.000Z",
      "releaseNotes": null,
      "version": "5.6.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": undefined,
      "releaseNotes": null,
      "version": "5.5.0",
    },
    Object {
      "changes": Array [],
      "compare": Object {},
      "date": "2018-08-24T14:23:00.000Z",
      "releaseNotes": null,
      "version": "5.4.0",
    },
  ],
}
`;
